### ISSUE TITLE
Failed to submit a form with file attachment

### ISSUE THREAD

**Issue by tranduyhung** [getgrav/grav#968]
I've tried to build a contact form with file attachment, after selecting a file and submitting the form, I get this error: "in_array() expects parameter 2 to be array, integer given". Here is the stackstrace

```
Whoops\Exception\ErrorException thrown with message "in_array() expects parameter 2 to be array, integer given"

Stacktrace:
#20 Whoops\Exception\ErrorException in system/src/Grav/Common/Data/Validation.php:42
#19 in_array in system/src/Grav/Common/Data/Validation.php:42
#18 Grav\Common\Data\Validation:validate in system/src/Grav/Common/Data/BlueprintSchema.php:77
#17 Grav\Common\Data\BlueprintSchema:validateArray in system/src/Grav/Common/Data/BlueprintSchema.php:38
#16 Grav\Common\Data\BlueprintSchema:validate in system/src/Grav/Common/Data/Blueprint.php:97
#15 Grav\Common\Data\Blueprint:validate in system/src/Grav/Common/Data/Data.php:191
#14 Grav\Common\Data\Data:validate in user/plugins/form/classes/form.php:273
#13 Grav\Plugin\Form:post in user/plugins/form/form.php:78
#12 Grav\Plugin\FormPlugin:onPageInitialized in vendor/symfony/event-dispatcher/EventDispatcher.php:184
#11 call_user_func in vendor/symfony/event-dispatcher/EventDispatcher.php:184
#10 Symfony\Component\EventDispatcher\EventDispatcher:doDispatch in vendor/symfony/event-dispatcher/EventDispatcher.php:46
#9 Symfony\Component\EventDispatcher\EventDispatcher:dispatch in vendor/rockettheme/toolbox/Event/src/EventDispatcher.php:23
#8 RocketTheme\Toolbox\Event\EventDispatcher:dispatch in system/src/Grav/Common/Grav.php:297
#7 Grav\Common\Grav:fireEvent in system/src/Grav/Common/Processors/PagesProcessor.php:19
#6 Grav\Common\Processors\PagesProcessor:process in system/src/Grav/Common/Grav.php:121
#5 Grav\Common\Grav:Grav\Common\{closure} in system/src/Grav/Common/Grav.php:383
#4 Grav\Common\Grav:Grav\Common\{closure} in system/src/Grav/Common/Grav.php:359
#3 call_user_func_array in system/src/Grav/Common/Grav.php:359
#2 Grav\Common\Grav:__call in system/src/Grav/Common/Grav.php:122
#1 Grav\Common\Grav:measureTime in system/src/Grav/Common/Grav.php:122
#0 Grav\Common\Grav:process in index.php:45
```

This is where the error occurs:

```
        // Special case for files, value is never empty and errors with code 4 instead.
        if (empty($validate['required']) && $field['type'] == 'file' && isset($value['error'])
                && ($value['error'] == UPLOAD_ERR_NO_FILE || in_array(UPLOAD_ERR_NO_FILE, $value['error']))) {
            return $messages;
        }
```

Here is the content of $value:

```
Array
(
    [name] => lorem
    [type] => application/octet-stream
    [tmp_name] => /tmp/phprxnZgg
    [error] => 0
    [size] => 6043
) 
```

As the error message already explains, the expected value of $value['error'] is an array, but in my case it is an integer which is equal to UPLOAD_ERR_OK.

I have tried to change the line 42 and 106

from

```
&& ($value['error'] == UPLOAD_ERR_NO_FILE || in_array(UPLOAD_ERR_NO_FILE, $value['error']))) {
```

to

```
&& ($value['error'] == UPLOAD_ERR_NO_FILE
|| (is_array($value['error']) && in_array(UPLOAD_ERR_NO_FILE, $value['error'])))) {
```

The I've got new error "Illegal string offset 'file'" for line 71 of user/plugins/email/email.php

Here is what from line 67 to 74:

```
                    $filesToAttach = (array)$params['attachments'];
                    if ($filesToAttach) foreach ($filesToAttach as $fileToAttach) {
                        $filesValues = $form->value($fileToAttach);
                        if ($filesValues) foreach($filesValues as $fileValues) {
                            $filename = $fileValues['file'];
                            $message->attach(\Swift_Attachment::fromPath($filename));
                        }
                    }
```

It seems the plugin expects $filesValues to be an array of files but in my case it is a single file?

```
Array
(
    [name] => lorem
    [type] => application/octet-stream
    [tmp_name] => /tmp/phpq3Unf7
    [error] => 0
    [size] => 6043
)
```
Just in case you need, here is my file upload field:

```
        -
            name: attachment
            label: 'Attachment'
            type: file
            multiple: false
            destination: 'user/data/attachments'
            accept:
                - application/pdf
                - application/x-pdf
                - application/zip
```

And here is the "process" section of the form:

```
        -
            email:
                from: '{{ config.plugins.email.from }}'
                to: ['{{ config.plugins.email.from }}']
                subject: '[Contact] {{ form.value.name|e }}'
                body: '{% include ''forms/data.html.twig'' %}'
                attachments:
                  - 'attachment'
```

Was there anyone having this problem before? Is this really a bug or something wrong from my side? Thanks!

**Comment by flaviocopes** [comment]
Can you paste the whole form definition, for quicker testing?

**Comment by tranduyhung** [comment]
Here is the MD file:

```
---
title: Contact
form:
    action: /contact
    name: contact-form
    fields:
        -
            name: name
            label: Name
            placeholder: 'Name *'
            type: text
            classes: pure-u-1-1
            validate:
                required: true
        -
            name: email
            label: Email
            placeholder: 'Email *'
            type: email
            classes: pure-u-1-1
            validate:
                required: true
        -
            name: phone
            label: Phone
            placeholder: 'Phone *'
            type: text
            classes: pure-u-1-1
            validate:
                required: true
        -
            name: message
            label: Message
            placeholder: 'Message *'
            type: textarea
            classes: pure-u-1-1
            validate:
                required: true
        -
            name: attachment
            label: 'Attachment'
            type: file
            multiple: false
            destination: 'user/data/attachments'
            accept:
                - application/pdf
                - application/x-pdf
                - application/zip
    buttons:
        -
            type: submit
            value: Send
            classes: 'pure-button pure-button-primary button-large '
    process:
        -
            email:
                from: '{{ config.plugins.email.from }}'
                to: ['{{ config.plugins.email.from }}']
                subject: '[Contact] {{ form.value.name|e }}'
                body: '{% include ''forms/data.html.twig'' %}'
                attachments:
                  - 'attachment'
        -
            save:
                fileprefix: contact-
                dateformat: Ymd-His-u
                extension: txt
                body: '{% include ''forms/data.txt.twig'' %}'
        -
            display: thank-you
---
```

**Comment by flaviocopes** [comment]
I tested it, I have no issues. 

Make sure you have the latest Grav 1.1 and Form plugin installed, is this the case?

**Comment by tranduyhung** [comment]
Yes, everything is in the latest version. Let me try again or try on a different server/host. If the problem still exists maybe I will upload the site or user folder to Github to make it easier to test. Thanks!

**Comment by flaviocopes** [comment]
Sure, or PM it to me on https://gitter.im/flaviocopes

**Comment by tranduyhung** [comment]
Thank you for your time helping me! I figured out the problem, the form twig file of my theme didn't have the line

```
{% set scope = scope ?: 'data.' %}
```

So the fields were not in "data" array and the attachment couldn't be found by Swift.

**Comment by zanami** [comment]
@tranduyhung Could you please clarify where did you put that "set scope"?
I'm experience the same problem - attachments don't get saved or sent, nothing happens.

**Comment by tranduyhung** [comment]
@zanami That line is in the default form.html.twig of Form plugin, I overrided the file to build my own form and accidentally deleted that line.

**Comment by zanami** [comment]
Thanks @tuphamnguyen 
I don't have form.html.twig so I guess the default is used
Form version: 2.5.0 Still doesn't work. 
I guess it's irrelevant to your issue

**Comment by flaviocopes** [comment]
@zanami just fixed and issued a new release. Was a problem introduced by 2.5.0
